src/main/java/de/tum/cit/aet/artemis/programming/service/tokens/UserTokenExpiryNotificationService.java

```diff
@@ -0,0 +1,64 @@
+package de.tum.cit.aet.artemis.programming.service.tokens;
+
+import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_SCHEDULING;
+import static java.time.ZonedDateTime.now;
+
+import java.time.ZonedDateTime;
+import java.util.function.Function;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import de.tum.cit.aet.artemis.communication.service.notifications.SingleUserNotificationService;
+import de.tum.cit.aet.artemis.core.domain.User;
+import de.tum.cit.aet.artemis.core.repository.UserRepository;
+
+@Profile(PROFILE_SCHEDULING)
+@Service
+public class UserTokenExpiryNotificationService {
+
+    private static final Logger log = LoggerFactory.getLogger(UserTokenExpiryNotificationService.class);
+
+    private final SingleUserNotificationService singleUserNotificationService;
+
+    private final UserRepository userRepository;
+
+    public UserTokenExpiryNotificationService(SingleUserNotificationService singleUserNotificationService, UserRepository userRepository) {
+        this.singleUserNotificationService = singleUserNotificationService;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * Schedules VCS access token expiry notifications to users every morning at 6:00:00 am
+     */
+    @Scheduled(cron = "0 0 6 * * *")
+    public void sendTokenExpirationNotifications() {
```

Comment:
I am confused, why do we notify users before their SSH key expires, but for tokens, we only do that after the token was expired? Would it not make sense to notify token users as well before hand so they can react to it like in the case of SSH?

Reply:
I agree, this was an oversight on my side somehow.
I added the notification to notify users on expiring tokens one week before that.